torch/distributed/fsdp/_init_utils.py

```diff
@@ -75,13 +88,140 @@
 @no_type_check
 def _init_process_group_state(
     state: _FSDPState,
-    process_group: Optional[dist.ProcessGroup],
+    process_group: ProcessGroupType,
+    sharding_strategy: ShardingStrategy,
+    policy: Optional[_FSDPPolicy],
 ) -> _FSDPState:
-    state.process_group = process_group or _get_default_group()
+    if sharding_strategy in HYBRID_SHARDING_STRATEGIES:
+        if process_group is None and policy is None:
+            # Raise an error here, since this is manual wrapping with no process group
+            # passed in, there is no way to ensure all wrapped FSDP units use the same
+            # process groups.
+            raise ValueError(
+                f"Manual wrapping with {sharding_strategy} requires explicit specification of process group."
+            )
+        else:
+            state = _init_process_group_state_for_hybrid_shard(state, process_group)
+            assert state.process_group is not None, "Expected to populate state.process_group for hybrid shard"
```

Comment:
Do we really want these asserts? This seems like we can just read `_init_process_group_state_for_hybrid_shard` to make sure these are set.

Reply:
[Not blocking] But please think about if we need/want these `assert`s

Reply:
I think these are useful, it's to make sure that `_init_process_group_state_for_hybrid_shard` sets all of these attributes.

In the future, `_init_process_group_state_for_hybrid_shard` could possibly be refactored or modified to set these attributes in helper functions, then reading through that would become a bit more cumbersome.

Reply:
I think the asserts should be at the end of `_init_process_group_state_for_hybrid_shard` then, representing a post-condition. This is just my personal design preference and probably does not matter in this case. However, in general, you would have to enforce the post-condition upon each call to `_init_process_group_state_for_hybrid_shard`.

(This is also how I approach post-conditions and invariants in general -- they should be coupled to the method/function itself, not their usages.)